base/sparse/sparsematrix.jl

```diff
@@ -1433,7 +1433,7 @@ function _broadcast_unary_nz2z_z2z_T{TvA,TiA,TvB}(f::Function, A::SparseMatrixCS
     return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
 end
 function _broadcast_unary_nz2z_z2z{Tv}(f::Function, A::SparseMatrixCSC{Tv})
-    _broadcast_unary_nz2z_z2z_T(f, A, Tv)
+    _broadcast_unary_nz2z_z2z_T(f, A, promote_op(f, Tv))
```

Comment:
@stevengj, I benchmarked two versions of your suggestion (for nz2nz_z2z-class unary operations). The overhead from calling `broadcast(f, view(A.nzval, 1:nnz(A)))` (and to a lesser degree `broadcast(f, A.nzval)`) seems substantial for small to moderate size sparse matrices. Thoughts? (Benchmarks below). Best!
```julia
function oldbc{Tv}(f::Function, A::SparseMatrixCSC{Tv})
    oldbc_T(f, A, Base.promote_op(f, Tv))
end
function oldbc_T{TvA,TiA,TvB}(f::Function, A::SparseMatrixCSC{TvA,TiA}, ::Type{TvB})
    Bcolptr = Vector{TiA}(A.n + 1)
    Browval = Vector{TiA}(nnz(A))
    Bnzval = Vector{TvB}(nnz(A))
    copy!(Bcolptr, 1, A.colptr, 1, A.n + 1)
    copy!(Browval, 1, A.rowval, 1, nnz(A))
    @inbounds @simd for k in 1:nnz(A)
        Bnzval[k] = f(A.nzval[k])
    end
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end

function newbc{TvA,TiA}(f::Function, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr = Vector{TiA}(A.n + 1)
    Browval = Vector{TiA}(nnz(A))
    copy!(Bcolptr, 1, A.colptr, 1, A.n + 1)
    copy!(Browval, 1, A.rowval, 1, nnz(A))
    Bnzval = broadcast(f, view(A.nzval, 1:nnz(A)))
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end

function newbcnv{TvA,TiA}(f::Function, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr = Vector{TiA}(A.n + 1)
    Browval = Vector{TiA}(nnz(A))
    copy!(Bcolptr, 1, A.colptr, 1, A.n + 1)
    copy!(Browval, 1, A.rowval, 1, nnz(A))
    Bnzval = broadcast(f, A.nzval)
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end

using BenchmarkTools

for N in [10^2, 10^3, 10^4]
    println(N)
    A = sprand(N, N, 0.1)
    oldbench = @benchmark oldbc($abs, $A)
    newbench = @benchmark newbc($abs, $A)
    newnvbench = @benchmark newbcnv($abs, $A)
    println(ratio(median(newnvbench), median(oldbench)))
    println(ratio(median(newbench), median(oldbench)))
end
```
yields
```julia
100
BenchmarkTools.TrialRatio:
  time:             11.641124140719773
  gctime:           1.0
  memory:           1.0632790028763184
  allocs:           7.75
  time tolerance:   5.00%
  memory tolerance: 1.00%
BenchmarkTools.TrialRatio:
  time:             11.815406389001213
  gctime:           1.0
  memory:           1.0661553211888783
  allocs:           8.0
  time tolerance:   5.00%
  memory tolerance: 1.00%
1000
BenchmarkTools.TrialRatio:
  time:             1.2436132534507298
  gctime:           1.0
  memory:           1.0006739279095351
  allocs:           5.833333333333333
  time tolerance:   5.00%
  memory tolerance: 1.00%
BenchmarkTools.TrialRatio:
  time:             1.6859576533933556
  gctime:           1.0
  memory:           1.000703660023191
  allocs:           6.0
  time tolerance:   5.00%
  memory tolerance: 1.00%
10000
BenchmarkTools.TrialRatio:
  time:             1.0039543258604076
  gctime:           1.0021435474746039
  memory:           1.0000067974835716
  allocs:           5.142857142857143
  time tolerance:   5.00%
  memory tolerance: 1.00%
BenchmarkTools.TrialRatio:
  time:             1.2308898483951354
  gctime:           1.004561220662931
  memory:           1.0000070973725528
  allocs:           5.285714285714286
  time tolerance:   5.00%
  memory tolerance: 1.00%
```